<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Installing PICRUSt &#8212; PICRUSt 1.1.3 documentation</title>
    
    <link rel="stylesheet" href="_static/sphinxdoc.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.1.3',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="nav-item nav-item-0"><a href="index.html">PICRUSt 1.1.3 documentation</a> &#187;</li> 
      </ul>
    </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Installing PICRUSt</a><ul>
<li><a class="reference internal" href="#recommended-install-picrust-with-bioconda">Recommended: Install PICRUSt with bioconda</a></li>
<li><a class="reference internal" href="#install-from-source-step-1-install-requirements">Install From Source Step 1. Install Requirements</a></li>
<li><a class="reference internal" href="#install-from-source-step-2-download-picrust">Install From Source Step 2. Download PICRUSt</a><ul>
<li><a class="reference internal" href="#release-software">Release software</a></li>
<li><a class="reference internal" href="#development-software">Development software</a></li>
</ul>
</li>
<li><a class="reference internal" href="#install-from-source-step-3-install-picrust">Install From Source Step 3. Install PICRUSt</a></li>
<li><a class="reference internal" href="#install-from-source-step-4-download-picrust-s-precalculated-files">Install from Source Step 4. Download PICRUSt&#8217;s precalculated files</a></li>
</ul>
</li>
</ul>

  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/install.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="installing-picrust">
<span id="install"></span><h1>Installing PICRUSt<a class="headerlink" href="#installing-picrust" title="Permalink to this headline">¶</a></h1>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Most users will not need to install PICRUSt, but can instead use the <a class="reference external" href="http://galaxy.morganlangille.com/">online Galaxy version</a>.</p>
</div>
<p>PICRUSt is written in python, and has been tested on Mac OS X and Linux systems. To install PICRUSt, first install all of the mandatory requirements, following the instructions on their respective websites. You should then download PICRUSt. You have the choice of downloading either the release version (recommended for most users) or the development version (recommended for users who need access to the latest features, and are willing to tolerate some instability in the code). Next, you will download the large precalculated PICRUSt files and place them in your picrust/data directory. Finally, you&#8217;ll install PICRUSt. Each of these steps are detailed below.</p>
<div class="section" id="recommended-install-picrust-with-bioconda">
<h2>Recommended: Install PICRUSt with bioconda<a class="headerlink" href="#recommended-install-picrust-with-bioconda" title="Permalink to this headline">¶</a></h2>
<p>The easiest way to install PICRUSt is with <a class="reference external" href="https://bioconda.github.io/">bioconda</a>.</p>
<p>To do so you need to type this command in a conda environment:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">conda</span> <span class="n">install</span> <span class="o">-</span><span class="n">c</span> <span class="n">bioconda</span> <span class="n">picrust</span>
</pre></div>
</div>
<p>You will then need to download the required precalculated files with this command (see more details under Step 4 below):</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">download_picrust_files</span><span class="o">.</span><span class="n">py</span>
</pre></div>
</div>
<p>You should be good to go! If you run into any problems you can try installing the source manually instead.</p>
</div>
<div class="section" id="install-from-source-step-1-install-requirements">
<h2>Install From Source Step 1. Install Requirements<a class="headerlink" href="#install-from-source-step-1-install-requirements" title="Permalink to this headline">¶</a></h2>
<p>Follow the install instructions found on the website of each of the dependencies below to install PICRUSt&#8217;s dependencies.</p>
<p><strong>Mandatory</strong></p>
<p>These dependencies are automatically installed with <cite>pip install .</cite> below:</p>
<ul class="simple">
<li><a class="reference external" href="http://www.python.org/">python</a> (version 2.7)</li>
<li><a class="reference external" href="http://www.pycogent.org/">PyCogent</a> (version 1.5.3)</li>
<li><a class="reference external" href="http://biom-format.org">biom</a> (version 2.1.6)</li>
</ul>
<p>h5py needs to be installed manually (numpy is also required, but it should be installed automatically with this package):</p>
<ul class="simple">
<li><a class="reference external" href="http://www.h5py.org">h5py</a> (version 2.7.0)</li>
</ul>
<p>You can install h5py with this command:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">pip</span> <span class="n">install</span> <span class="n">h5py</span>
</pre></div>
</div>
<p><strong>Rebuilding PICRUSt&#8217;s precalculated 16S rRNA OR Genome Predictions (optional)</strong></p>
<ul class="simple">
<li><a class="reference external" href="http://www.r-project.org/">R</a> installed with <a class="reference external" href="http://ape.mpl.ird.fr/">APE</a> library</li>
</ul>
</div>
<div class="section" id="install-from-source-step-2-download-picrust">
<h2>Install From Source Step 2. Download PICRUSt<a class="headerlink" href="#install-from-source-step-2-download-picrust" title="Permalink to this headline">¶</a></h2>
<div class="section" id="release-software">
<h3>Release software<a class="headerlink" href="#release-software" title="Permalink to this headline">¶</a></h3>
<p>The latest release of PICRUSt is <a class="reference external" href="https://github.com/picrust/picrust/releases/download/v1.1.3/picrust-1.1.3.tar.gz">picrust-1.1.3</a>.</p>
<p>We recommend the release version of PICRUSt for most users. If you&#8217;re not sure whether you want the release or the development version of PICRUSt, you should likely go with the release version.</p>
</div>
<div class="section" id="development-software">
<h3>Development software<a class="headerlink" href="#development-software" title="Permalink to this headline">¶</a></h3>
<p>Alternatively you can download the latest development version of PICRUSt. You can download the development version of PICRUSt from <a class="reference external" href="https://github.com/picrust/picrust/archive/master.zip">this link</a>, or using the following command if you have <code class="docutils literal"><span class="pre">git</span></code> installed:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">git</span> <span class="n">clone</span> <span class="n">git</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">picrust</span><span class="o">/</span><span class="n">picrust</span><span class="o">.</span><span class="n">git</span> <span class="n">picrust</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="install-from-source-step-3-install-picrust">
<h2>Install From Source Step 3. Install PICRUSt<a class="headerlink" href="#install-from-source-step-3-install-picrust" title="Permalink to this headline">¶</a></h2>
<p>After downloading PICRUSt, you&#8217;ll need to unzip the file. If you&#8217;ve downloaded the release version, do this with the following command:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">tar</span> <span class="o">-</span><span class="n">xzf</span> <span class="n">picrust</span><span class="o">-</span><span class="mf">1.1</span><span class="o">.</span><span class="mf">3.</span><span class="n">tar</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
<p>You&#8217;ll then change into the new <code class="docutils literal"><span class="pre">picrust-1.1.3</span></code> directory as follows:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">cd</span> <span class="n">picrust</span><span class="o">-</span><span class="mf">1.1</span><span class="o">.</span><span class="mi">3</span>
</pre></div>
</div>
<p>And finally, you&#8217;ll install PICRUSt with the following command:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">pip</span> <span class="n">install</span> <span class="o">.</span>
</pre></div>
</div>
</div>
<div class="section" id="install-from-source-step-4-download-picrust-s-precalculated-files">
<h2>Install from Source Step 4. Download PICRUSt&#8217;s precalculated files<a class="headerlink" href="#install-from-source-step-4-download-picrust-s-precalculated-files" title="Permalink to this headline">¶</a></h2>
<p>PICRUSt precomputes most of the computationally intensive pipeline so each user can get predictions with less steps. These files are changed whenever there is a new release of the GreenGenes tree and/or a new release of IMG. These files are fairly large and need to be downloaded separately. Also, you only need the version that corresponds to the GreenGenes that you picked OTUs against (see <a class="reference internal" href="tutorials/otu_picking.html#otu-picking-tutorial"><span class="std std-ref">Picking OTUs for use in PICRUSt</span></a>).</p>
<p>Assuming you will be picking OTUs against the newest version of GreenGenes and want KEGG Ortholog predictions, then the default files will be sufficient. If you picked against an older version of GreenGenes or are interested in different functional predictions (e.g. COGs) then see the complete list of <a class="reference internal" href="picrust_precalculated_files.html#id1"><span class="std std-ref">PICRUSt&#8217;s Precalculated Files</span></a>.</p>
<p>You can install the default files automatically by running:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">download_picrust_files</span><span class="o">.</span><span class="n">py</span>
</pre></div>
</div>
<p>Alternatively, download these files and place them in the <code class="docutils literal"><span class="pre">data</span></code> directory where PICRUSt was installed (e.g. if installed in a conda enviroment named &#8220;picrust&#8221;, place the files in <code class="docutils literal"><span class="pre">$HOME/miniconda3/envs/picrust/lib/python2.7/site-packages/picrust/data/</span></code>):</p>
<ul class="simple">
<li>Download for <a class="reference external" href="http://kronos.pharmacology.dal.ca/public_files/picrust/picrust_precalculated_v1.1.3/13_5/16S_13_5_precalculated.tab.gz">16S copy number normalization</a></li>
<li>Download for <a class="reference external" href="http://kronos.pharmacology.dal.ca/public_files/picrust/picrust_precalculated_v1.1.3/13_5/ko_13_5_precalculated.tab.gz">KEGG Ortholog predictions</a></li>
</ul>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="nav-item nav-item-0"><a href="index.html">PICRUSt 1.1.3 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2011-2013, The PICRUSt Project.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.5.6.
    </div>
  </body>
</html>