.. _citing_picrust:

Citing PICRUSt
===============

On this page we’ve compiled both the PICRUSt citation and links to several tools that PICRUSt is built on to make citing these various software packages easier.

**Predictive functional profiling of microbial communities using 16S rRNA marker gene sequences.** Langille, M. G.I.\*; Zaneveld, J.\*; Caporaso, J. G.; McDonald, D.; Knights, D.; a Reyes, J.; Clemente, J. C.; Burkepile, D. E.; Vega Thurber, R. L.; Knight, R.; Beiko, R. G.; and Huttenhower, C. *Nature Biotechnology*, 1-10. 8 2013. 

The manuscript describing PICRUSt can be found `here <http://www.nature.com/nbt/journal/vaop/ncurrent/abs/nbt.2676.html>`_


Citing ancestral state reconstruction algorithms used by PICRUSt
================================================================
PICRUSt is based on several tools for ancestral state reconstruction.   We would greatly appreciate your citation of these tools as well when possible. If you used the precalculated table of gene counts (this is the default), or ran ancestral_state_reconstruction.py using the ace_pic, ace_ml, or ace_reml tools:

**APE: Analyses of phylogenetics and evolution in R language.** 
Paradis, E.; Claude, J. & Strimmer, K. *Bioinformatics* **20**, 289–290 (2004).

If you used the parsimony criterion in ancestral_state_reconstruction.py (not typical), you may wish to cite the COUNT package: 

**Count: evolutionary analysis of phylogenetic profiles with parsimony and likelihood.** Csuros, M. *Bioinformatics* **26**, 1910–1912 (2010).


Citing tools used to generate your OTU table
============================================
If you used QIIME to generate your reference-picked OTU table (recommended), then this `page <http://www.wernerlab.org/software/macqiime/citations>`_ maintained by the Werner lab  has an excellent description of papers associated with the various tools that are part of the pipeline.  Although this can end up being a lot of citations, giving a shout out to tool developers is a great thing for the community.


