

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Pre-PI-CRUST from the Huttenhower Lab &mdash; PI-CRUST 0.1 documentation</title>
    
    <link rel="stylesheet" href="../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="top" title="PI-CRUST 0.1 documentation" href="../index.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li><a href="../index.html">PI-CRUST 0.1 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="pre-pi-crust-from-the-huttenhower-lab">
<span id="josh"></span><h1>Pre-PI-CRUST from the Huttenhower Lab<a class="headerlink" href="#pre-pi-crust-from-the-huttenhower-lab" title="Permalink to this headline">¶</a></h1>
<p>This document covers the code in the <tt class="docutils literal"><span class="pre">pre_picrust/huttenhower/</span></tt> directory. Questions on usage should be directed to Joshua Reyes (<a class="reference external" href="mailto:jreyes&#37;&#52;&#48;post&#46;harvard&#46;edu">jreyes<span>&#64;</span>post<span>&#46;</span>harvard<span>&#46;</span>edu</a>).</p>
<div class="section" id="description-of-code">
<h2>Description of code<a class="headerlink" href="#description-of-code" title="Permalink to this headline">¶</a></h2>
<dl class="docutils">
<dt>This code performs the following functions that are the backbone of PI-CRUST:</dt>
<dd><ul class="first last simple">
<li>Constructs a (NCBI) labeled phylogeny from the input OTU abundance table.</li>
<li>Performs ancestral state reconstruction from the tree using a decaying exponential weighted vote to produce a gene count for each missing genome.</li>
<li>Predicts functional abundances simply by multiplying OTU abundance by gene count and summing across all OTUs.</li>
</ul>
</dd>
</dl>
</div>
<div class="section" id="software-requirements">
<h2>Software Requirements<a class="headerlink" href="#software-requirements" title="Permalink to this headline">¶</a></h2>
<blockquote>
<div><ul class="simple">
<li>Python &gt;= 2.6</li>
<li>Numpy, a package for scientific computing in Python</li>
<li>ETE 2, a package for the manipulation, analysis, and visualization of phylogenetic trees</li>
<li>SCons, a software construction tool similar to the Make utility</li>
</ul>
</div></blockquote>
<p>Additionally, I like to use:</p>
<blockquote>
<div><ul class="simple">
<li>virtualenv, a tool to create isolated Python environments</li>
<li>virtualenvwrapper, syntactic sugar for virtualenv</li>
<li>pip, a package manager for Python</li>
</ul>
</div></blockquote>
</div>
<div class="section" id="installation">
<h2>Installation<a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h2>
<p>PICRUST itself doesn&#8217;t require an installation. Simply copy our code into a directory $PICRUST_HOME. To get things running, though, you need to install the packages PICRUST depends on.</p>
<div class="section" id="step-1-create-an-isolated-python-environment">
<h3>Step 1: Create an isolated Python environment<a class="headerlink" href="#step-1-create-an-isolated-python-environment" title="Permalink to this headline">¶</a></h3>
<p>Set up a Python environment specifically to run PICRUST. This keeps the packages you install for PICRUST separate from all of your other projects:</p>
<div class="highlight-python"><pre>mkvirtualenv picrust</pre>
</div>
</div>
<div class="section" id="step-2-install-the-dependencies">
<h3>Step 2: Install the dependencies<a class="headerlink" href="#step-2-install-the-dependencies" title="Permalink to this headline">¶</a></h3>
<p>To install the dependencies, I like to use pip.</p>
<p>I&#8217;ve included a copy of SCons 2.1.0 and ETE2 2.0 rev 111 in the etc/ folder for your convenience. You can rely on pip to install Numpy.</p>
<div class="highlight-python"><pre>pip install numpy
cd $PICRUST_HOME/etc
tar xzfv scons-2.1.0.tar.gz
tar xzfv ete2-2.0rev111.tar.gz
cd scons-2.1.0
python setup.py install
cd ../ete2-2.0rev111
python setup.py install</pre>
</div>
<p>PICRUST doesn&#8217;t use the database or visualization ETE toolkits. The ETE setup script will warn you that you won&#8217;t be able to use them unless you install more packages. That&#8217;s okay, though. Just answer yes to continue with its installation.</p>
<div class="highlight-python"><pre>Installing ETE (A python Environment for Tree Exploration).

Checking dependencies...
MySQLdb cannot be found in your python installation.
MySQLdb is required for the PhylomeDB access API.
PyQt4 cannot be found in your python installation.
PyQt4 is required for tree visualization and rendering.

However, you can still install ETE without such funtionalites.
Do you want to continue with the installation? [y,n] y</pre>
</div>
</div>
<div class="section" id="step-3-make-anfunc-path-accessible">
<h3>Step 3: Make ANFunc path accessible<a class="headerlink" href="#step-3-make-anfunc-path-accessible" title="Permalink to this headline">¶</a></h3>
<p>PICRUST relies on an a binary file called ANFunc. This executable isn&#8217;t actually part of PIECRUST; it&#8217;s part of another project in the lab. It was written in C++ to multiply matrices fast. It can probably be replaced with something in SciPy later on.</p>
<p>For example:</p>
<div class="highlight-python"><pre>echo "export PATH=$PICRUST_HOME/bin/:$PATH" &gt;&gt; ~/.bashrc</pre>
</div>
</div>
</div>
<div class="section" id="running-picrust">
<h2>Running PICRUST<a class="headerlink" href="#running-picrust" title="Permalink to this headline">¶</a></h2>
<div class="section" id="input">
<h3>Input<a class="headerlink" href="#input" title="Permalink to this headline">¶</a></h3>
<p>Simply drop as many OTU abundance PCL files as you like into the $PICRUST/input/ directory. Two other files, a phylogeny of taxa <tt class="docutils literal"><span class="pre">$PICRUST/taxdump.txt</span></tt> and KO table <tt class="docutils literal"><span class="pre">genomes_02-zro-flt.txt</span></tt> should also be included. You can find the KO table in the Dropbox:</p>
<div class="highlight-python"><pre>cp utility_data_files/huttenhower_prepicrust_input/* pre_picrust/huttenhower/input</pre>
</div>
</div>
<div class="section" id="running">
<h3>Running<a class="headerlink" href="#running" title="Permalink to this headline">¶</a></h3>
<p>SCons drives the entire pipeline, which makes for a simple execution. Once everything is in place in the input directory, run <tt class="docutils literal"><span class="pre">scons</span></tt> from $PICRUST_HOME. The <tt class="docutils literal"><span class="pre">SConstruct</span></tt> file directs <tt class="docutils literal"><span class="pre">scons</span></tt> and can be modified to suit your needs.</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">scons</span>
</pre></div>
</div>
</div>
<div class="section" id="output">
<h3>Output<a class="headerlink" href="#output" title="Permalink to this headline">¶</a></h3>
<p>Final results are saved in <tt class="docutils literal"><span class="pre">output/&lt;input</span> <span class="pre">name&gt;_kos.pcl</span></tt> by default.</p>
<p>Intermediate calculations are saved in <tt class="docutils literal"><span class="pre">output</span></tt> folder. Trees are saved using an internal representation in <tt class="docutils literal"><span class="pre">.fir</span></tt> files. Intermediate tables have end in the suffix <tt class="docutils literal"><span class="pre">_fir.pcl</span></tt>.</p>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Pre-PI-CRUST from the Huttenhower Lab</a><ul>
<li><a class="reference internal" href="#description-of-code">Description of code</a></li>
<li><a class="reference internal" href="#software-requirements">Software Requirements</a></li>
<li><a class="reference internal" href="#installation">Installation</a><ul>
<li><a class="reference internal" href="#step-1-create-an-isolated-python-environment">Step 1: Create an isolated Python environment</a></li>
<li><a class="reference internal" href="#step-2-install-the-dependencies">Step 2: Install the dependencies</a></li>
<li><a class="reference internal" href="#step-3-make-anfunc-path-accessible">Step 3: Make ANFunc path accessible</a></li>
</ul>
</li>
<li><a class="reference internal" href="#running-picrust">Running PICRUST</a><ul>
<li><a class="reference internal" href="#input">Input</a></li>
<li><a class="reference internal" href="#running">Running</a></li>
<li><a class="reference internal" href="#output">Output</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="../_sources/prepicrust/huttenhower.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li><a href="../index.html">PI-CRUST 0.1 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2011, Joshua Reyes, Greg Caporaso.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.2pre.
    </div>
  </body>
</html>