

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>PI-CRUST ala Morgan written in Perl &mdash; PI-CRUST 0.1 documentation</title>
    
    <link rel="stylesheet" href="../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="top" title="PI-CRUST 0.1 documentation" href="../index.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li><a href="../index.html">PI-CRUST 0.1 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="pi-crust-ala-morgan-written-in-perl">
<span id="morgan"></span><h1>PI-CRUST ala Morgan written in Perl<a class="headerlink" href="#pi-crust-ala-morgan-written-in-perl" title="Permalink to this headline">¶</a></h1>
<p>This document covers the code in the <tt class="docutils literal"><span class="pre">pre_picrust/morgan/</span></tt> directory. Questions on usage should be directed to Morgan Langille (<a class="reference external" href="mailto:morgan&#46;g&#46;i&#46;langille&#37;&#52;&#48;gmail&#46;com">morgan<span>&#46;</span>g<span>&#46;</span>i<span>&#46;</span>langille<span>&#64;</span>gmail<span>&#46;</span>com</a>)</p>
<div class="section" id="description-of-code">
<h2>Description of code<a class="headerlink" href="#description-of-code" title="Permalink to this headline">¶</a></h2>
<dl class="docutils">
<dt>This code performs the following functions that are the backbone of PICRUST:</dt>
<dd><ul class="first last simple">
<li>Creates a reference 16S tree for sequences from reference genomes using PyNAST and RaXML (done once for a given set of genomes; &#8220;offline&#8221;).</li>
<li>Performs ancestral state reconstruction (ASR) using various methods and various functional databases for the reference tree (done once per reference genome; &#8220;offline&#8221;).</li>
<li>Takes raw metagenomic reads, aligns and filters them using pynast, and places them onto a tree using pplacer. (done once for each metagenome set; &#8220;online&#8221;)</li>
<li>Predicts functional abundances for the set of reads in the metagenome sample using a variety of methods including ASR, Nearest Neighbour, and Random.</li>
</ul>
</dd>
<dt>There is also code to do &#8220;leave one out&#8221; validation using only reference genomes:</dt>
<dd><ul class="first last simple">
<li>Can calculate accuracy (precision and recall) for any of the prediction methods on a genome basis for many functional databases (incl. PFAM, SEED, EC).</li>
<li>Simiarly, calculate accuracy across the various functional classes within a functional database</li>
<li>Create numerous scatter plots, violin (distribution) plots, and bar charts.</li>
</ul>
</dd>
</dl>
</div>
<div class="section" id="software-requirements">
<h2>Software Requirements<a class="headerlink" href="#software-requirements" title="Permalink to this headline">¶</a></h2>
<blockquote>
<div><ul class="simple">
<li>Perl &gt;= v 5.10</li>
<li>Perl Modules (BioPerl, Log::Log4perl, JSON, Parallel::ForkManager)</li>
<li>PyNAST and GreenGenes template (included with QIIME installation)</li>
<li>R</li>
<li>R Modules (ape,geiger,vioplot, sciplot)</li>
</ul>
</div></blockquote>
</div>
<div class="section" id="installation">
<h2>Installation<a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h2>
<blockquote>
<div><p>Data resource files are not included with the code, but need to be copied from the shared Dropbox folder:</p>
<div class="highlight-python"><pre>cp genome_test_cases/seed_EC_and_pfam pre_picrust/morgan/data</pre>
</div>
<p>That&#8217;s it! You are ready to run PI-CRUST.</p>
</div></blockquote>
</div>
<div class="section" id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<blockquote>
<div><ul class="simple">
<li>The main scripts needed to run the pipeline are in the main directory.</li>
<li>These main scripts call other scripts within bin/ which can be accessed but are not well documented.</li>
<li>To get documentation on any of the main scripts you can run the script with the &#8211;help option to get the full documentation. (e.g. build_asr.pl &#8211;help).</li>
<li>Reference trees and other &#8220;offline&#8221; files will be stored in a sub-directory of the directory &#8220;ref_trees&#8221;. The name of the sub-directory will be named according to the file given to build the reference tree (see usage below for build_ref_16s.pl)</li>
<li>Metagenomic read alignments, predictions, and other &#8220;online&#8221; files will be stored in a sub-directory of &#8220;tmp&#8221;. The name of the sub-directory will be named according to the file containing the metagenomic reads.</li>
<li>&#8220;run_leave_one_out.pl&#8221; is a pipeline script that glues together all of the steps and runs it on a single reference genome leaving that genome out of the reference dataset and treating it&#8217;s 16S sequence as a single metagenomic read. This script also calls other additional scripts in /bin to test accuracy and create plots.</li>
</ul>
</div></blockquote>
</div>
<div class="section" id="usage-examples">
<h2>Usage examples<a class="headerlink" href="#usage-examples" title="Permalink to this headline">¶</a></h2>
<p>This section provides a description of how to run the main code of the PI-CRUST pipeline:</p>
<div class="section" id="step-1-create-offline-files-for-completed-genomes">
<h3>Step 1: Create &#8220;offline&#8221; files for completed genomes<a class="headerlink" href="#step-1-create-offline-files-for-completed-genomes" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="step-1a-create-a-reference-16s-tree-for-completed-genomes">
<h3>Step 1a: Create a reference 16S tree for completed genomes<a class="headerlink" href="#step-1a-create-a-reference-16s-tree-for-completed-genomes" title="Permalink to this headline">¶</a></h3>
<p>Here&#8217;s how to create a reference tree from your genome 16S sequences:</p>
<div class="highlight-python"><pre>build_ref_16s.pl your_genomes_16S.fa</pre>
</div>
<p>This takes an input fasta file of 16S sequences from completed genomes. IDs of in the fasta file will be used as genome identifiers.
Output is written to the directory <tt class="docutils literal"><span class="pre">ref_trees/your_genomes_16S.fa</span></tt>.
Various <tt class="docutils literal"><span class="pre">pynast_*</span></tt> and <tt class="docutils literal"><span class="pre">RAxML_*</span></tt> files are created.</p>
</div>
<div class="section" id="step-1b-create-ancestral-state-reconstructions-asrs-for-internal-nodes-of-reference-tree">
<h3>Step 1b: Create ancestral state reconstructions (ASRs) for internal nodes of reference tree<a class="headerlink" href="#step-1b-create-ancestral-state-reconstructions-asrs-for-internal-nodes-of-reference-tree" title="Permalink to this headline">¶</a></h3>
<p>You&#8217;ll next need to run <tt class="docutils literal"><span class="pre">build_asr.pl</span></tt> on the <tt class="docutils literal"><span class="pre">your_genomes_16S.fa</span></tt> from step 1:</p>
<div class="highlight-python"><pre>build_asr.pl -f pfam -m pic -r your_genomes_16S.fa</pre>
</div>
<p>This performs the users choice of ASR method (pic, ML, REML, etc.) on the reference tree using the user&#8217;s choice of functional traits (pfam, subsystem, EC, etc.).
Output is written to the directory <tt class="docutils literal"><span class="pre">ref_trees/your_genomes_16S.fa</span></tt>.
Another version of the reference tree, but with internal node labels added, will created as <tt class="docutils literal"><span class="pre">RAxML_result.16s_with_node_labels</span></tt>.
ASR information will be stored in a compressed tab delimited table (<tt class="docutils literal"><span class="pre">&lt;func&gt;_&lt;method&gt;_counts.txt.gz</span></tt>) where column labels correspond to newly created internal tree labels and row labels are function labels.</p>
</div>
<div class="section" id="step-2-make-predictions-for-metagenomic-reads">
<h3>Step 2: Make predictions for metagenomic reads<a class="headerlink" href="#step-2-make-predictions-for-metagenomic-reads" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="step-2a-place-metagenomic-reads-onto-reference-tree">
<h3>Step 2a: Place metagenomic reads onto reference tree<a class="headerlink" href="#step-2a-place-metagenomic-reads-onto-reference-tree" title="Permalink to this headline">¶</a></h3>
<p>The metagenomic reads need to be filtered (by pynast) and placed onto the reference tree (by pplacer):</p>
<div class="highlight-python"><pre>place_reads.pl -r your_genomes_16S.fa meta_reads.fa</pre>
</div>
<p>This takes the reference tree created in Step 1 and places raw metagenomic 16S reads onto the tree.
Output is written to the directory <tt class="docutils literal"><span class="pre">tmp/meta_reads.fa</span></tt>.
<tt class="docutils literal"><span class="pre">pynast_trimmed_alignment.jplace</span></tt> and <tt class="docutils literal"><span class="pre">pplacer.tog.tree</span></tt> are json and tree representations,respectively, of the placement output from pplacer.</p>
</div>
<div class="section" id="step-2b-make-predictions-for-the-metagenomic-reads">
<h3>Step 2b: Make predictions for the metagenomic reads<a class="headerlink" href="#step-2b-make-predictions-for-the-metagenomic-reads" title="Permalink to this headline">¶</a></h3>
<p>Make actual functional abundance predictions for each of the functions can be specified:</p>
<div class="highlight-python"><pre>make_predictions.pl -f pfam -m pic -r your_genomes_16S.fa -q meta_reads.fa</pre>
</div>
<p>The command requires the reference tree (created in Step 1 above) and the metagenomic reads name (same as used in step 1a).
The user can choose which functional trait (pfam, subsystem, EC, etc.) and the method (pic, ML, neighbour, random).
Predictions are stored in the directory <tt class="docutils literal"><span class="pre">tmp/meta_reads.fa</span></tt> in the file <tt class="docutils literal"><span class="pre">&lt;func&gt;_&lt;method&gt;_predictions.txt</span></tt> (e.g <tt class="docutils literal"><span class="pre">pfam_pic_predictions.txt</span></tt>).</p>
</div>
</div>
<div class="section" id="other-usage">
<h2>Other Usage<a class="headerlink" href="#other-usage" title="Permalink to this headline">¶</a></h2>
<div class="section" id="validate-accuracy-of-pi-crust">
<h3>Validate accuracy of PI-CRUST<a class="headerlink" href="#validate-accuracy-of-pi-crust" title="Permalink to this headline">¶</a></h3>
<p>Runs the entire validation pipeline including the creation of figures:</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">run_leave_one_out</span><span class="o">.</span><span class="n">pl</span> <span class="o">-</span><span class="n">r</span><span class="o">=</span><span class="n">your_genomes_16S</span><span class="o">.</span><span class="n">fa</span>
</pre></div>
</div>
<p>Calculates the accuracy of the PI-CRUST software using various functional databases and difference methods for prediction (ASR, NN, random).
The pipeline removes one genome from the reference genome dataset and treats it as a unknown query (metagenomic) 16S sequence.
The pipeline makes predictions on this genome and the predicted functional abundances are compared to the known annotated functional abundances for that genome.
This is repeated for every reference genome in the dataset (&gt;1000 times), thus is very CPU intensive. Some stages such as reference tree construction and ASR may want to be done in pieces on a cluster. Also, if the user has a multi-core machine they can use the &#8211;threads option.</p>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">PI-CRUST ala Morgan written in Perl</a><ul>
<li><a class="reference internal" href="#description-of-code">Description of code</a></li>
<li><a class="reference internal" href="#software-requirements">Software Requirements</a></li>
<li><a class="reference internal" href="#installation">Installation</a></li>
<li><a class="reference internal" href="#overview">Overview</a></li>
<li><a class="reference internal" href="#usage-examples">Usage examples</a><ul>
<li><a class="reference internal" href="#step-1-create-offline-files-for-completed-genomes">Step 1: Create &#8220;offline&#8221; files for completed genomes</a></li>
<li><a class="reference internal" href="#step-1a-create-a-reference-16s-tree-for-completed-genomes">Step 1a: Create a reference 16S tree for completed genomes</a></li>
<li><a class="reference internal" href="#step-1b-create-ancestral-state-reconstructions-asrs-for-internal-nodes-of-reference-tree">Step 1b: Create ancestral state reconstructions (ASRs) for internal nodes of reference tree</a></li>
<li><a class="reference internal" href="#step-2-make-predictions-for-metagenomic-reads">Step 2: Make predictions for metagenomic reads</a></li>
<li><a class="reference internal" href="#step-2a-place-metagenomic-reads-onto-reference-tree">Step 2a: Place metagenomic reads onto reference tree</a></li>
<li><a class="reference internal" href="#step-2b-make-predictions-for-the-metagenomic-reads">Step 2b: Make predictions for the metagenomic reads</a></li>
</ul>
</li>
<li><a class="reference internal" href="#other-usage">Other Usage</a><ul>
<li><a class="reference internal" href="#validate-accuracy-of-pi-crust">Validate accuracy of PI-CRUST</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="../_sources/prepicrust/morgan.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li><a href="../index.html">PI-CRUST 0.1 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2011, Joshua Reyes, Greg Caporaso.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.2pre.
    </div>
  </body>
</html>