

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Picking OTUs for use in PICRUSt &mdash; PICRUSt 0.9.2 documentation</title>
    
    <link rel="stylesheet" href="../_static/default.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '0.9.2',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="top" title="PICRUSt 0.9.2 documentation" href="../index.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li><a href="../index.html">PICRUSt 0.9.2 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="picking-otus-for-use-in-picrust">
<span id="otu-picking-tutorial"></span><h1>Picking OTUs for use in PICRUSt<a class="headerlink" href="#picking-otus-for-use-in-picrust" title="Permalink to this headline">¶</a></h1>
<div class="section" id="introduction">
<h2>Introduction<a class="headerlink" href="#introduction" title="Permalink to this headline">¶</a></h2>
<p>This document covers how to pick OTUs from marker gene data to use with PICRUSt. To do this, you&#8217;ll use a &#8216;closed-reference&#8217; OTU picking protocol where you search sequences against the GG reference OTUs at a specified percent identity, and discard any reads that don&#8217;t hit that reference collection. The newest available reference collection can be found here:</p>
<blockquote>
<div><ul class="simple">
<li>gg_13_5_otus.tar.gz (<a class="reference external" href="http://greengenes.secondgenome.com/downloads/database/13_5">download</a> )</li>
</ul>
</div></blockquote>
<p>This tutorial assumes that you have QIIME installed. See the <a class="reference external" href="http://www.qiime.org">QIIME website</a> for details on how to use QIIME. The quickest way to get started with QIIME is working on the Amazon Web Services cloud, and you can find instructions for <a class="reference external" href="http://qiime.org/tutorials/working_with_aws.html">using QIIME on the cloud here</a>.</p>
</div>
<div class="section" id="picking-closed-reference-otus-with-qiime">
<h2>Picking closed reference OTUs with QIIME<a class="headerlink" href="#picking-closed-reference-otus-with-qiime" title="Permalink to this headline">¶</a></h2>
<p>To pick &#8216;closed reference&#8217; OTUs with QIIME for use in PICRUST, you should begin with a demuliplexed fasta file in QIIME format, and the IMG/GG reference collection.</p>
<p>The demuliplexed fasta file in QIIME format is a standard multi-line fasta file, where sequence identifiers are of the form <tt class="docutils literal"><span class="pre">sampleID_seqID</span></tt>. In these sequence identifiers, <tt class="docutils literal"><span class="pre">sampleID</span></tt> indicates the name of the sample, and <tt class="docutils literal"><span class="pre">seqID</span></tt> is a unique (with respect to the current file) sequence identifier. The <tt class="docutils literal"><span class="pre">seqID</span></tt> values are often just assigned as ascending integers to ensure their uniqueness. These files can be generated with the <tt class="docutils literal"><span class="pre">split_libraries</span></tt> * scripts in QIIME, or using other software. This file might look like:</p>
<div class="highlight-python"><pre>&gt;s1_1
ACCTTAGGATGGATTAGACCCAGA
&gt;s1_2
ACCAGGATGGACCCCTTAGACCCAGA
&gt;s2_3
AGGTCCTGGATGGACCCCTTAGACCCAGA
&gt;s1_4
ACCAGTTGATGGACCCCTTAGACCCAGA</pre>
</div>
<p>In this example we have two samples (<tt class="docutils literal"><span class="pre">s1</span></tt> and <tt class="docutils literal"><span class="pre">s2</span></tt>) and four sequences, three of which are in <tt class="docutils literal"><span class="pre">s1</span></tt> and one of which is in <tt class="docutils literal"><span class="pre">s2</span></tt>. For additional details on this file, see <a class="reference external" href="http://qiime.org/documentation/file_formats.html#demultiplexed-sequences">here</a>.</p>
<p>If your demultiplexed fna file is named <tt class="docutils literal"><span class="pre">seqs.fna</span></tt> and is in your current working directory, and you have unzipped the GG OTUs in the current working directory, you would pick OTUs with the following commands:</p>
<div class="highlight-python"><pre>echo "pick_otus:enable_rev_strand_match True"  &gt;&gt; $PWD/otu_picking_params_97.txt
echo "pick_otus:similarity 0.97" &gt;&gt; $PWD/otu_picking_params_97.txt
pick_closed_reference_otus.py -i $PWD/seqs.fna -o $PWD/ucrC97/ -p $PWD/otu_picking_params_97.txt -r $PWD/gg_13_5_otus/rep_set/97_otus.fasta</pre>
</div>
<p>This command picks OTUs and builds a <a class="reference external" href="http://www.biom-format.org">biom-formatted OTU table</a>, with OTUs assigned at 97% identity. The primary file of interest will be <tt class="docutils literal"><span class="pre">ucrC97/uclust_ref_picked_otus/otu_table.biom</span></tt>, which will be the OTU table that you pass to PICRUSt.</p>
<p>Alternatively, if you&#8217;d like to pick OTUs at a 90% percent identity, you could run the following command:</p>
<div class="highlight-python"><pre>echo "pick_otus:enable_rev_strand_match True"  &gt;&gt; otu_picking_params_90.txt
echo "pick_otus:similarity 0.90" &gt;&gt; otu_picking_params_90.txt
pick_closed_reference_otus.py -i $PWD/seqs.fna -o $PWD/ucrC90/ -p $PWD/otu_picking_params_90.txt -r $PWD/gg_13_5_otus/rep_set/97_otus.fasta</pre>
</div>
<p>If you have parallel QIIME running in your environment, you can modify the commands to run in parallel by append the <tt class="docutils literal"><span class="pre">-a</span></tt> flag and the <tt class="docutils literal"><span class="pre">-O</span></tt> parameter. This might looking like the following to use 8 processors to pick OTUs at 97% identity:</p>
<div class="highlight-python"><pre>echo "pick_otus:enable_rev_strand_match True"  &gt;&gt; $PWD/otu_picking_params_97.txt
echo "pick_otus:similarity 0.97" &gt;&gt; $PWD/otu_picking_params_97.txt
pick_closed_reference_otus.py -i $PWD/seqs.fna -o $PWD/ucrC97/ -p $PWD/otu_picking_params_97.txt -r $PWD/gg_13_5_otus/rep_set/97_otus.fasta -a -O 8</pre>
</div>
<p>Note that you should not specify a value for <tt class="docutils literal"><span class="pre">-O</span></tt> that is greater than the number of processor/cores that you have available. This will result in longer run times as multiple jobs will compete for the same processor.</p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Picking OTUs for use in PICRUSt</a><ul>
<li><a class="reference internal" href="#introduction">Introduction</a></li>
<li><a class="reference internal" href="#picking-closed-reference-otus-with-qiime">Picking closed reference OTUs with QIIME</a></li>
</ul>
</li>
</ul>

  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="../_sources/tutorials/otu_picking.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li><a href="../index.html">PICRUSt 0.9.2 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2011-2013, The PICRUSt Project.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.1.3.
    </div>
  </body>
</html>