<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>categorize_by_function.py – Collapse table data to a specified level in a hierarchy. &#8212; PICRUSt 1.1.3 documentation</title>
    
    <link rel="stylesheet" href="../_static/sphinxdoc.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.1.3',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="nav-item nav-item-0"><a href="../index.html">PICRUSt 1.1.3 documentation</a> &#187;</li> 
      </ul>
    </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/scripts/categorize_by_function.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <span class="target" id="categorize-by-function"></span><div class="section" id="categorize-by-function-py-collapse-table-data-to-a-specified-level-in-a-hierarchy">
<span id="index-0"></span><h1><em>categorize_by_function.py</em> &#8211; Collapse table data to a specified level in a hierarchy.<a class="headerlink" href="#categorize-by-function-py-collapse-table-data-to-a-specified-level-in-a-hierarchy" title="Permalink to this headline">¶</a></h1>
<p><strong>Description:</strong></p>
<p>This script collapses hierarchical data to a specified level. For instance, often it is useful to examine KEGG results from a higher level within the pathway hierarchy. Many genes are sometimes involved in multiple pathways, and in these circumstances (also know as a one-to-many relationship), the gene is counted for each pathway. This has a side effect of increasing the total count of genes in the table.</p>
<p><strong>Usage:</strong> <code class="file docutils literal"><span class="pre">categorize_by_function.py</span> <span class="pre">[options]</span></code></p>
<p><strong>Input Arguments:</strong></p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p><strong>[REQUIRED]</strong></p>
<dl class="docutils">
<dt>-i, <cite>-</cite>-input_fp</dt>
<dd>The predicted metagenome table</dd>
<dt>-o, <cite>-</cite>-output_fp</dt>
<dd>The resulting table</dd>
<dt>-c, <cite>-</cite>-metadata_category</dt>
<dd>The metadata category that describes the hierarchy</dd>
<dt>-l, <cite>-</cite>-level</dt>
<dd>The level in the hierarchy to collapse to. A value of 0 is not allowed, a value of 1 is the highest level, and any higher value nears the leaves of the hierarchy. For instance, if the hierarchy contains 4 levels, specifying 3 would collapse at one level above being fully specified.</dd>
</dl>
<p><strong>[OPTIONAL]</strong></p>
<dl class="last docutils">
<dt><cite>-</cite>-ignore</dt>
<dd>Ignore the comma separated list of names. For instance, specifying &#8211;ignore_unknown=unknown,unclassified will ignore those labels while collapsing. The default is to not ignore anything. [default: None]</dd>
<dt>-f, <cite>-</cite>-format_tab_delimited</dt>
<dd>Output the predicted metagenome table in tab-delimited format [default: False]</dd>
</dl>
</div>
<p><strong>Output:</strong></p>
<p>Output table is contains gene counts at a higher level within a hierarchy.</p>
<p>Collapse predicted metagenome using KEGG Pathway metadata.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">categorize_by_function</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">i</span> <span class="n">predicted_metagenomes</span><span class="o">.</span><span class="n">biom</span> <span class="o">-</span><span class="n">c</span> <span class="n">KEGG_Pathways</span> <span class="o">-</span><span class="n">l</span> <span class="mi">3</span> <span class="o">-</span><span class="n">o</span> <span class="n">predicted_metagenomes</span><span class="o">.</span><span class="n">L3</span><span class="o">.</span><span class="n">biom</span>
</pre></div>
</div>
<p>Change output to tab-delimited format (instead of BIOM).</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">categorize_by_function</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">f</span> <span class="o">-</span><span class="n">i</span> <span class="n">predicted_metagenomes</span><span class="o">.</span><span class="n">biom</span> <span class="o">-</span><span class="n">c</span> <span class="n">KEGG_Pathways</span> <span class="o">-</span><span class="n">l</span> <span class="mi">3</span> <span class="o">-</span><span class="n">o</span> <span class="n">predicted_metagenomes</span><span class="o">.</span><span class="n">L3</span><span class="o">.</span><span class="n">txt</span>
</pre></div>
</div>
<p>Collapse predicted metagenome using taxonomy metadata (not one-to-many).</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">categorize_by_function</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">i</span> <span class="n">observation_table</span><span class="o">.</span><span class="n">biom</span> <span class="o">-</span><span class="n">c</span> <span class="n">taxonomy</span> <span class="o">-</span><span class="n">l</span> <span class="mi">1</span> <span class="o">-</span><span class="n">o</span> <span class="n">observation_table</span><span class="o">.</span><span class="n">L1</span><span class="o">.</span><span class="n">biom</span>
</pre></div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="nav-item nav-item-0"><a href="../index.html">PICRUSt 1.1.3 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2011-2013, The PICRUSt Project.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.5.6.
    </div>
  </body>
</html>