<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>predict_metagenomes.py – This script produces the actual metagenome functional predictions for a given OTU table. &#8212; PICRUSt 1.1.3 documentation</title>
    
    <link rel="stylesheet" href="../_static/sphinxdoc.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.1.3',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="nav-item nav-item-0"><a href="../index.html">PICRUSt 1.1.3 documentation</a> &#187;</li> 
      </ul>
    </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/scripts/predict_metagenomes.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <span class="target" id="predict-metagenomes"></span><div class="section" id="predict-metagenomes-py-this-script-produces-the-actual-metagenome-functional-predictions-for-a-given-otu-table">
<span id="index-0"></span><h1><em>predict_metagenomes.py</em> &#8211; This script produces the actual metagenome functional predictions for a given OTU table.<a class="headerlink" href="#predict-metagenomes-py-this-script-produces-the-actual-metagenome-functional-predictions-for-a-given-otu-table" title="Permalink to this headline">¶</a></h1>
<p><strong>Description:</strong></p>
<p><strong>Usage:</strong> <code class="file docutils literal"><span class="pre">predict_metagenomes.py</span> <span class="pre">[options]</span></code></p>
<p><strong>Input Arguments:</strong></p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p><strong>[REQUIRED]</strong></p>
<dl class="docutils">
<dt>-i, <cite>-</cite>-input_otu_table</dt>
<dd>The input otu table in biom format</dd>
<dt>-o, <cite>-</cite>-output_metagenome_table</dt>
<dd>The output file for the predicted metagenome</dd>
</dl>
<p><strong>[OPTIONAL]</strong></p>
<dl class="last docutils">
<dt>-t, <cite>-</cite>-type_of_prediction</dt>
<dd>Type of functional predictions. Valid choices are: ko, cog, rfam [default: ko]</dd>
<dt>-g, <cite>-</cite>-gg_version</dt>
<dd>Version of GreenGenes that was used for OTU picking. Valid choices are: 13_5, 18may2012 [default: 13_5]</dd>
<dt>-c, <cite>-</cite>-input_count_table</dt>
<dd>Precalculated function predictions on per otu basis in biom format (can be gzipped). Note: using this option overrides &#8211;type_of_prediction and &#8211;gg_version. [default: None]</dd>
<dt>-a, <cite>-</cite>-accuracy_metrics</dt>
<dd>If provided, calculate accuracy metrics for the predicted metagenome.  NOTE: requires that per-genome accuracy metrics were calculated using <a class="reference external" href="./predict_traits.html">predict_traits.py</a> during genome prediction (e.g. there are &#8220;NSTI&#8221; values in the genome .biom file metadata)</dd>
<dt><cite>-</cite>-suppress_subset_loading</dt>
<dd>Normally, only counts for OTUs present in the sample are loaded.  If this flag is passed, the full biom table is loaded.  This makes no difference for the analysis, but may result in faster load times (at the cost of more memory usage)</dd>
<dt><cite>-</cite>-load_precalc_file_in_biom</dt>
<dd>Instead of loading the precalculated file in tab-delimited format (with otu ids as row ids and traits as columns) load the data in biom format (with otu as SampleIds and traits as ObservationIds) [default: False]</dd>
<dt><cite>-</cite>-input_variance_table</dt>
<dd>Precalculated table of variances corresponding to the precalculated table of function predictions.  As with the count table, these are on a per otu basis and in BIOM format (can be gzipped). Note: using this option overrides &#8211;type_of_prediction and &#8211;gg_version. [default: None]</dd>
<dt><cite>-</cite>-with_confidence</dt>
<dd>Calculate 95% confidence intervals for metagenome predictions.  By default, this uses the confidence intervals for the precalculated table of genes for greengenes OTUs.  If you pass a custom count table with -c and select this option, you must also specify a corresponding table of confidence intervals for the gene content prediction using &#8211;input_variance_table. (these are generated by running <a class="reference external" href="./predict_traits.html">predict_traits.py</a> with the &#8211;with_confidence option). If this flag is set, three addtional output files will be generated, named the same as the metagenome prediction output, but with .variance .upper_CI or .lower_CI appended immediately before the file extension[default: False]</dd>
<dt>-f, <cite>-</cite>-format_tab_delimited</dt>
<dd>Output the predicted metagenome table in tab-delimited format [default: False]</dd>
</dl>
</div>
<p><strong>Output:</strong></p>
<p>Output is a table of function counts (e.g. KEGG KOs) by sample ids.</p>
<p>Predict KO abundances for a given OTU table picked against the newest version of GreenGenes.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">predict_metagenomes</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">i</span> <span class="n">normalized_otus</span><span class="o">.</span><span class="n">biom</span> <span class="o">-</span><span class="n">o</span> <span class="n">predicted_metagenomes</span><span class="o">.</span><span class="n">biom</span>
</pre></div>
</div>
<p>Change output format to plain tab-delimited:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">predict_metagenomes</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">f</span> <span class="o">-</span><span class="n">i</span> <span class="n">normalized_otus</span><span class="o">.</span><span class="n">biom</span> <span class="o">-</span><span class="n">o</span> <span class="n">predicted_metagenomes</span><span class="o">.</span><span class="n">txt</span>
</pre></div>
</div>
<p>Predict COG abundances for a given OTU table.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">predict_metagenomes</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">i</span> <span class="n">normalized_otus</span><span class="o">.</span><span class="n">biom</span> <span class="o">-</span><span class="n">t</span> <span class="n">cog</span> <span class="o">-</span><span class="n">o</span> <span class="n">cog_predicted_metagenomes</span><span class="o">.</span><span class="n">biom</span>
</pre></div>
</div>
<p>Output confidence intervals for each prediction.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">predict_metagenomes</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">i</span> <span class="n">normalized_otus</span><span class="o">.</span><span class="n">biom</span> <span class="o">-</span><span class="n">o</span> <span class="n">predicted_metagenomes</span><span class="o">.</span><span class="n">biom</span> <span class="o">--</span><span class="n">with_confidence</span>
</pre></div>
</div>
<p>Predict metagenomes using a custom trait table in tab-delimited format.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">predict_metagenomes</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">i</span> <span class="n">otu_table_for_custom_trait_table</span><span class="o">.</span><span class="n">biom</span> <span class="o">-</span><span class="n">c</span> <span class="n">custom_trait_table</span><span class="o">.</span><span class="n">tab</span> <span class="o">-</span><span class="n">o</span> <span class="n">output_metagenome_from_custom_trait_table</span><span class="o">.</span><span class="n">biom</span>
</pre></div>
</div>
<p>Predict metagenomes,variances,and 95% confidence intervals for each gene category using a custom trait table in tab-delimited format.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">predict_metagenomes</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">i</span> <span class="n">otu_table_for_custom_trait_table</span><span class="o">.</span><span class="n">biom</span> <span class="o">--</span><span class="n">input_variance_table</span> <span class="n">custom_trait_table_variances</span><span class="o">.</span><span class="n">tab</span> <span class="o">-</span><span class="n">c</span> <span class="n">custom_trait_table</span><span class="o">.</span><span class="n">tab</span> <span class="o">-</span><span class="n">o</span> <span class="n">output_metagenome_from_custom_trait_table</span><span class="o">.</span><span class="n">biom</span> <span class="o">--</span><span class="n">with_confidence</span>
</pre></div>
</div>
<p>Change the version of GG used to pick OTUs</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="n">predict_metagenomes</span><span class="o">.</span><span class="n">py</span> <span class="o">-</span><span class="n">i</span> <span class="n">normalized_otus</span><span class="o">.</span><span class="n">biom</span> <span class="o">-</span><span class="n">g</span> <span class="mi">18</span><span class="n">may2012</span> <span class="o">-</span><span class="n">o</span> <span class="n">predicted_metagenomes</span><span class="o">.</span><span class="n">biom</span>
</pre></div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="nav-item nav-item-0"><a href="../index.html">PICRUSt 1.1.3 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2011-2013, The PICRUSt Project.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.5.6.
    </div>
  </body>
</html>